grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1083,12 +1087,109 @@ class StockMovementService {
+            csvMapReader.fieldKeys = [
+                    'id',
+                    'code',
+                    'name',
+                    'lot',
+                    'expiration',
+                    'binLocation',
+                    'quantity',
+            ]
+            return csvMapReader.toList()
+
+        } catch (Exception e) {
+            throw new RuntimeException("Error parsing order item CSV: " + e.message, e)
+        }
+    }
+
+    void validatePicklistListImport(List<Map> data, List<PickPageItem> pickPageItems, Boolean supportsOverPick = false) {
+        Map<String, List> groupedItems = data.groupBy { it.id }
+
+        data?.eachWithIndex { Map params, index ->
+            params.errors = new ArrayList()
+            if (!params.id) {
+                params.errors.push("Requisition item id is required")
+            }
+            if (params.quantity == null) {
+                params.errors.push("Quantity picked are required")
+            }
+            if (params.lot?.contains("E") && NumberUtils.isNumber(params.lot)) {
+                params.errors.push("Lot numbers must not be specified in scientific notation. " +
+                        "Please reformat field with Lot Number: \"${params.lot}\" to a number format")
+            }
+
+            PickPageItem pickPageItem = pickPageItems.find { it.requisitionItem?.id == params.id }
+            if (!pickPageItem) {
+                params.errors.push("Requisition item id: ${params.id} not found")
+            }
+            if (pickPageItem) {
+                AvailableItem availableItem = pickPageItem.availableItems?.find { item ->
+                    Boolean binLocationMatches = params.binLocation ? item.binLocation?.name == params.binLocation : !item.binLocation
+                    Boolean lotMatches = params.lot == item.inventoryItem?.lotNumber
+                    binLocationMatches && lotMatches
+                }
+
+                if (!availableItem) {
+                    params.errors.push("There is no item available with: ${params.lot ? "lot ${params.lot}" : ""} ${params.binLocation ? "bin ${params.binLocation}" : ""}")
+                }
+            }
+
+            Integer itemQuantitySum = groupedItems[params.id].sum { Integer.parseInt(it.quantity) }
+            if (pickPageItem && itemQuantitySum > pickPageItem.requisitionItem.quantity) {
+                if (!supportsOverPick) {
+                    params.errors.push("Item ${params.id} is overpicked, expected quantity ${pickPageItem.requisitionItem.quantity}, received ${itemQuantitySum}")
+                }
+            } else if (itemQuantitySum != pickPageItem.requisitionItem.quantity) {
+                params.errors.push("Item ${params.id} expected quantity is ${pickPageItem.requisitionItem.quantity}, received ${itemQuantitySum}")
+            }
+        }
+    }
+
+    void importPicklistTemplate(List<Map> data, StockMovement stockMovement, List<PickPageItem> pickPageItems) {
```

Comment:
This is nit-picky but let's make the following changes:
* change the method name since we're not importing a template, we're importing picklist items from a template
* we don't really need StockMovement, so let's try to pass only the args that are needed to the method
* make the List<Map> data the last argument 
* let's change the name of the data argument be more clear
* consider converting from Map to a Command class, if desired

```
void importPicklistItems(Picklist picklist, List<PickPageItem> pickPageItems, 
    List<Map> picklistItems) 
```
Passing pickPageItems here is bothering me but I don't see a good solution at the moment. Maybe if it looked like this (with pickPageItems being included with the stock movement) 
```
void importPicklistItems(StockMovement stockMovement, List<Map> picklistItems)
```
... then it'd look cleaner. But I don't like the idea of passing a huge command object like StockMovement to a method that only needs access to a few objects under it.

It's possible that we could move all of this into a DTO/Command class hierarchy and that would have only what is necessary to do the work 
```
void importPicklistItems(ImportPicklistCommand command) { 
    ...
}

class ImportPicklistCommand { 
    Picklist picklist
    Location currentLocation
    StockMovement stockMovement
    List<PicklistItemCommand> picklistItems
}

class ImportPicklistItemCommand {
    RequisitionItem requisitionItem
    List<AvailableItem> availableItems
    // Include this if preferred, but the command should only reference the objects it 
    // needed to perform validation and make decisions
    //PickPageItem pickPageItem

    // move validation logic here
}
```